rate-limit/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+//! # Rate Limit
+//! A module to provide rate limit for arbitrary type Key and integer type Value
+//!
+//! - [`Config`](./trait.Config.html)
+//! - [`Call`](./enum.Call.html)
+//! - [`Module`](./struct.Module.html)
+//!
+//! ## Overview
+//!
+//! This module is a utility to provide rate limiter for arbitrary type Key and
+//! integer type Value, which can config limit rule to produce quota and consume
+//! quota, and expose quota consuming checking and whitelist that can bypass
+//! checks.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use codec::MaxEncodedLen;
+use frame_support::{pallet_prelude::*, traits::UnixTime, transactional, BoundedVec};
+use frame_system::pallet_prelude::*;
+use orml_traits::{RateLimiter, RateLimiterError};
+use orml_utilities::OrderedSet;
+use scale_info::TypeInfo;
+use sp_runtime::traits::{SaturatedConversion, Zero};
+use sp_std::{prelude::*, vec::Vec};
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+mod mock;
+mod tests;
+pub mod weights;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	/// Limit rules type.
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum RateLimitRule {
+		/// Each `blocks_count` blocks to reset remainer quota to `quota`
+		/// amount. is_allowed check return true when the remainer quota gte the
+		/// consume amount.
+		PerBlocks { blocks_count: u64, quota: u128 },
```

Comment:
maybe we could have a `enum Period { Blocks(u64), Seconds(u64) }` and able to merge `PerBlocks` and `PerSeconds` into `PerPeriod { period: Period, quota: u128 }`